### PR TITLE
UHM-5709 Show Exposed Infant form for 18 months old and less

### PR DESCRIPTION
**Pull request by brandones**
https://pihemr.atlassian.net/browse/UHM-5709

### CODE REVIEW

File: api/src/main/java/org/openmrs/module/mirebalais/require/RequireUtil.java [hunk 1]
```diff
@@ -71,6 +71,10 @@ public static String patientYoungerThan(int age) {
         return "patient.person.age <= " + age;
     }
 
+    public static String patientYoungerThanMonths(int ageMonths) {
+        return "(new Date(visit.startDatetime) - new Date(patient.person.birthdate)) / (1000 * 60 * 60 * 24 * 30.4) <= " + ageMonths;
```
Comment (mseaton) [inline api/src/main/java/org/openmrs/module/mirebalais/require/RequireUtil.java:75]: Can we try:  moment(visit.startDatetime).diff(moment(visit.patient.person.birthdate), 'months')

File: api/src/main/java/org/openmrs/module/mirebalais/require/RequireUtil.java [hunk 2]
```diff
@@ -71,6 +71,10 @@ public static String patientYoungerThan(int age) {
         return "patient.person.age <= " + age;
     }
 
+    public static String patientYoungerThanMonths(int ageMonths) {
```
Comment (mseaton) [inline api/src/main/java/org/openmrs/module/mirebalais/require/RequireUtil.java:74]: I have no idea why whoever implemented patientYoungerThan(int) decided to make that actually return "younger than or equal to", but let's not make that mistake again here.  If it's youngerThan, then use < ageMonths.  If it's youngerThanOrEqualTo, then name the method that.

I'd also be more explicit with the name like "patientYoungerThanMonthsAtVisitStart", so it's clear that this is relative to the visit start date, rather than the current date, especially since the other "patientYoungerThan" method doesn't take the visitStartdate into play at all.

Speaking of which...we should probably also change the patientYoungerThan method to factor in the visit start date, and ideally change it so that the method matches the behavior as discussed above.  @mogoodrich ?

File: api/src/main/java/org/openmrs/module/mirebalais/apploader/CustomAppLoaderFactory.java [hunk 3]
```diff
@@ -2304,7 +2304,7 @@ private void enableEIDForm() {
                 Privileges.TASK_EMR_ENTER_HIV_CONSULT_NOTE.privilege(),
                 and(sessionLocationHasTag("HIV Consult Location"),
                         visitDoesNotHaveEncounterOfType(PihEmrConfigConstants.ENCOUNTERTYPE_EID_FOLLOWUP_UUID),
-                        and(patientYoungerThan(2)),
+                        and(patientYoungerThanMonths(18)),
```
Comment (mseaton) [inline api/src/main/java/org/openmrs/module/mirebalais/apploader/CustomAppLoaderFactory.java:2307]: This should either be patientYoungerThanMonths(19) or patientYoungerThanOrEqualToMonths(18), depending on how you decide to change things below.

**Review by mseaton — CHANGES_REQUESTED** [review]

File: api/src/main/java/org/openmrs/module/mirebalais/require/RequireUtil.java (hunk 1 above)
Comment (mogoodrich) [reply api/src/main/java/org/openmrs/module/mirebalais/require/RequireUtil.java:75]: Worth a shot, but hese require rules are written in JavaScript to allow them to portable and evaluated either client or server side...  in Java, they are evaluated using the Nashorn engine within the Appframework, and assumedly moment is not available there (though we could potentially see about loading it see: AppFrameworkServiceImpl:99

File: api/src/main/java/org/openmrs/module/mirebalais/require/RequireUtil.java (hunk 2 above)
Comment (mogoodrich) [reply api/src/main/java/org/openmrs/module/mirebalais/require/RequireUtil.java:74]: We should fix the name of the patientYoungerThan method, and, yes, it should likely factor in the visit start date. (we can check to see where it is actually used, but it seems likely we want age at visit, not current age)

File: api/src/main/java/org/openmrs/module/mirebalais/require/RequireUtil.java (hunk 1 above)
Comment (brandones) [reply api/src/main/java/org/openmrs/module/mirebalais/require/RequireUtil.java:75]: Yeah, tried that, `moment` is not available.

File: api/src/main/java/org/openmrs/module/mirebalais/require/RequireUtil.java (hunk 2 above)
Comment (brandones) [reply api/src/main/java/org/openmrs/module/mirebalais/require/RequireUtil.java:73]: Fixed. I'm not sure if `AtVisitStart` is really useful, since I can't think of anything in the visit context that *isn't* relative to the visit/encounter time. The actual current date is almost never relevant.

Comment (mseaton) [reply api/src/main/java/org/openmrs/module/mirebalais/require/RequireUtil.java:73]: I won't push on the "AtVisitStart" - I don't really care that much.  But there could definitely be other dates of interest.  the current date _might_ be what is actually desired, for an in-patient stay longer than a day, or we might have use cases in the future where we want to have logic based around "ageAtProgramEnrollment" or something.  So this helps to be clear.  But I don't care enough to worry for now.

File: api/src/main/java/org/openmrs/module/mirebalais/require/RequireUtil.java [hunk 4]
```diff
@@ -66,9 +66,12 @@ public static String patientIsChild() {  // expects patient or visit to be in th
...
     }
 
-    // TODO: note this is really younger than or equal to...
     public static String patientYoungerThan(int age) {
-        return "patient.person.age <= " + age;
+        return "patient.person.age < " + age;
+    }
+
+    public static String patientYoungerThanMonths(int ageMonths) {
+        return "(new Date(visit.startDatetime) - new Date(patient.person.birthdate)) / (1000 * 60 * 60 * 24 * 30.4) < " + ageMonths;
```
Comment (mseaton) [inline api/src/main/java/org/openmrs/module/mirebalais/require/RequireUtil.java:74]: I do care enough about fixing this still.  If we can't use moment, that's fine, but we can build a method that is more correct.  Even if we don't think anyone will care about exact accuracy, I think we should still want it to be as accurate as possible.  What about something like this - can you run this through some tests and confirm that it is accurate for various edge case scenarios and change to it if it works?

```var d1 = new Date('2018-06-06');
var d2 = new Date('2021-05-01');

// Take the years difference and multiply by 12 months
// Then, take the months difference and add it.  This will increment or decrement from the above appropriately
// Then, if the day of the month of the future date has not reached the day of the month of the earlier date, subtract a month

var months = ((d2.getFullYear() - d1.getFullYear()) * 12) + d2.getMonth() - d1.getMonth() - (d2.getDate() < d1.getDate() ? 1 : 0);

console.log(months);```

**Review by mseaton — CHANGES_REQUESTED** [review]

File: api/src/main/java/org/openmrs/module/mirebalais/require/RequireUtil.java (hunk 4 above)
Comment (mogoodrich) [reply api/src/main/java/org/openmrs/module/mirebalais/require/RequireUtil.java:74]: I just glanced at this, but it might be worth looking into loading moment into the Nashorn context. It might not even be much more work than writing the custom code (and will pay off since we can continue to use it).

https://anuradha-15.medium.com/loading-scripts-using-nashorn-85585f495cf0

(I guess we'd also have to confirm that moment is available in any basic JavaScript context where we process the rules, but I assuming it most likely will be)

Comment (mogoodrich) [reply api/src/main/java/org/openmrs/module/mirebalais/require/RequireUtil.java:74]: https://github.com/openmrs/openmrs-module-appframework/blob/master/api/src/main/java/org/openmrs/module/appframework/service/AppFrameworkServiceImpl.java#L98

Comment (brandones) [reply api/src/main/java/org/openmrs/module/mirebalais/require/RequireUtil.java:74]: Nice, that sounds good @mogoodrich . Sorry I've got to ditch this ticket though—I'm off tomorrow and offline till Tuesday. Hopefully one of you can carry it forward.

Note that the error of the naive technique never exceeds 1.4 days:

Screenshot from 2021-09-02 18-45-53

Comment (mogoodrich) [reply api/src/main/java/org/openmrs/module/mirebalais/require/RequireUtil.java:74]: Interesting, nice analysis...
